ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/StateBackupService.java

```diff
@@ -0,0 +1,469 @@
+        bodyWriter.writeBytes(bodyOutput.encoded().toArrayUnsafe());
+
+        final BytesValueRLPOutput receiptsOutput = new BytesValueRLPOutput();
+        receiptsOutput.writeList(receipts.get(), TransactionReceipt::writeToWithRevertReason);
+        receiptsWriter.writeBytes(receiptsOutput.encoded().toArrayUnsafe());
+
+        backupStatus.storedBlock = blockNumber;
+      }
+    }
+  }
+
+  private TrieIterator.State visitAccountStorage(
+      final Bytes32 nodeKey, final Node<Bytes> node, final RollingFileWriter accountFileWriter) {
+    backupStatus.currentStorage = nodeKey;
+
+    final BytesValueRLPOutput output = new BytesValueRLPOutput();
+    output.startList();
+    output.writeBytes(nodeKey);
+    output.writeBytes(node.getValue().orElse(Bytes.EMPTY));
+    output.endList();
+
+    try {
+      accountFileWriter.writeBytes(output.encoded().toArrayUnsafe());
+    } catch (final IOException ioe) {
+      LOG.error("Failure writing backup", ioe);
+      return State.STOP;
+    }
+
+    backupStatus.storageCount.incrementAndGet();
+    return State.CONTINUE;
+  }
+
+  static class RollingFileWriter implements Closeable {
+    final BiFunction<Integer, Boolean, Path> filenameGenerator;
+    final boolean compressed;
+    int currentSize;
+    int fileNumber;
+    FileOutputStream out;
+    final DataOutputStream index;
+
+    RollingFileWriter(
+        final BiFunction<Integer, Boolean, Path> filenameGenerator, final boolean compressed)
+        throws FileNotFoundException {
+      this.filenameGenerator = filenameGenerator;
+      this.compressed = compressed;
+      currentSize = 0;
+      fileNumber = 0;
+      final Path firstOutputFile = filenameGenerator.apply(fileNumber, compressed);
+      out = new FileOutputStream(firstOutputFile.toFile());
+      index = new DataOutputStream(new FileOutputStream(dataFileToIndex(firstOutputFile).toFile()));
+    }
+
+    void writeBytes(final byte[] bytes) throws IOException {
+      final byte[] finalBytes;
+      if (compressed) {
+        finalBytes = Snappy.compress(bytes);
+      } else {
+        finalBytes = bytes;
+      }
+      int pos = currentSize;
+      currentSize += finalBytes.length;
+      if (currentSize > MAX_FILE_SIZE) {
+        out.close();
+        out = new FileOutputStream(filenameGenerator.apply(++fileNumber, compressed).toFile());
+        currentSize = finalBytes.length;
+        pos = 0;
+      }
+      index.writeShort(fileNumber);
+      index.writeInt(pos);
+      out.write(finalBytes);
+    }
+
+    @Override
+    public void close() throws IOException {
+      out.close();
+      index.close();
+    }
+  }
+
+  public static final class BackupStatus {
+    long targetBlock;
+    long storedBlock;
+    boolean compressed;
+    Bytes32 currentAccount;
+    Bytes32 currentStorage;
+    AtomicLong accountCount = new AtomicLong(0);
+    AtomicLong codeSize = new AtomicLong(0);
+    AtomicLong storageCount = new AtomicLong(0);
+    boolean requestAccepted;
```

Comment:
This is for the future looking JSON-RPC api.  It will work like the AdminGenerateLogBloomCache where if a backup is in progress the request to backup is rejected.